unprotected object

Fail on GLCI, likely caused by #3610 

```
Running test id 1682.6      
Running test id 1682.7      Test 1682.7 ran without errors but failed check that x equals y:
> x = dt1[dt3, on = .(x >= V1, y <= V2), sum(z) * V3, by = .EACHI] 
    x  y V1  [Key= Types=int,int,dou Classes=int,int,num]
1: 12 15 NA                                              
2:  5  7 84                                              
> y = dt1[dt3, on = .(x >= V1, y <= V2)][, sum(z) * V3[1L], by = .(x,      y)] 
First 6 of 1 (type 'character'): [1] "Error in bmerge(i, x, leftcols, rightcols, roll, rollends, nomatch, mult,  : \n  unprotected object (0x4b03c48) encountered (was INTSXP)\n"
Error in bmerge(i, x, leftcols, rightcols, roll, rollends, nomatch, mult,  : 
  unprotected object (0x4b03c48) encountered (was INTSXP)
Calls: test.data.table ... sys.source -> eval -> eval -> [ -> [.data.table -> bmerge
Execution halted
ERROR: Job failed: exit code 1
```

found this
```
fmelt.c: In function ‘getvaluecols’:
fmelt.c:487:53: warning: ‘ithisidx’ may be used uninitialized in this function [-Wmaybe-uninitialized]
             itarget[counter + k] = ithiscol[ithisidx[k]-1];
```

--- a/CRAN_Release.cmd
+++ b/CRAN_Release.cmd
@@ -92,14 +92,16 @@ grep mkChar *.c            # see comment in bmerge.c about passing this grep. I'
 # or ii) passed to setAttrib (which protects, providing leak-seals above are ok)
 # ScalarLogical in R now returns R's global TRUE from R 3.1.0; Apr 2014. Before that it allocated.
 # Aside: ScalarInteger may return globals for small integers in future version of R.
-grep ScalarInteger *.c   # Check all Scalar* either PROTECTed, return-ed or passed to setAttrib.
+grep ScalarInteger *.c | grep -v PROTECT | grep -v setAttrib | grep -v return  # Check all Scalar* either PROTECTed, return-ed or passed to setAttrib.
+grep ScalarString *.c  | grep -v PROTECT | grep -v setAttrib | grep -v return
 grep ScalarLogical *.c   # Now we depend on 3.1.0+, check ScalarLogical is NOT PROTECTed.
-grep ScalarString *.c
 
 # Inspect missing PROTECTs
 # To pass this grep is why we like SET_VECTOR_ELT(,,var=allocVector()) style on one line.
 # If a PROTECT is not needed then a comment is added explaining why and including "PROTECT" in the comment to pass this grep
 grep allocVector *.c | grep -v PROTECT | grep -v SET_VECTOR_ELT | grep -v setAttrib | grep -v return
+grep coerceVector *.c | grep -v PROTECT | grep -v SET_VECTOR_ELT | grep -v setAttrib | grep -v return
+grep asCharacter *.c | grep -v PROTECT | grep -v SET_VECTOR_ELT | grep -v setAttrib | grep -v return
 
 cd ..
 R

--- a/inst/tests/tests.Rraw
+++ b/inst/tests/tests.Rraw
@@ -6685,6 +6685,9 @@ test(1477.07, transpose(strsplit(ll, ",", fixed=TRUE)), tstrsplit(ll, ",", fixed
 test(1477.08, transpose(1:5), error="l must be a list")
 test(1477.09, transpose(list(as.complex(c(1, 1+5i)))), error="Unsupported column type")
 test(1477.10, transpose(list(list(1:5))), error="Item 1 of list input is")
+test(1477.11, transpose(as.list(1:5), fill=1:2), error="fill must be NULL or length=1 vector")
+test(1477.12, transpose(as.list(1:5), ignore.empty=NA), error="ignore.empty should be logical TRUE/FALSE")
+test(1477.13, transpose(list()), list())
 
 # #480 `setDT` and 'lapply'
 ll = list(data.frame(a=1), data.frame(x=1, y=2), NULL, list())

--- a/src/assign.c
+++ b/src/assign.c
@@ -279,7 +279,7 @@ SEXP assign(SEXP dt, SEXP rows, SEXP cols, SEXP newcolnames, SEXP values)
   // cols : column names or numbers corresponding to the values to set
   // rows : row numbers to assign
   R_len_t i, j, numToDo, targetlen, vlen, r, oldncol, oldtncol, coln, protecti=0, newcolnum, indexLength;
-  SEXP targetcol, RHS, names, nullint, thisv, targetlevels, newcol, s, colnam, klass, tmp, colorder, key, index, a, assignedNames, indexNames;
+  SEXP targetcol, names, nullint, thisv, targetlevels, newcol, s, colnam, klass, tmp, colorder, key, index, a, assignedNames, indexNames;
   SEXP bindingIsLocked = getAttrib(dt, install(".data.table.locked"));
   bool verbose=GetVerbose(), anytodelete=false, isDataTable=false;
   const char *c1, *tc1, *tc2;
@@ -327,7 +327,7 @@ SEXP assign(SEXP dt, SEXP rows, SEXP cols, SEXP newcolnames, SEXP values)
     // fast way to assign to whole column, without creating 1:nrow(x) vector up in R, or here in C
   } else {
     if (isReal(rows)) {
-      rows = PROTECT(rows = coerceVector(rows, INTSXP)); protecti++;
+      rows = PROTECT(coerceVector(rows, INTSXP)); protecti++;
       warning("Coerced i from numeric to integer. Please pass integer for efficiency; e.g., 2L rather than 2");
     }
     if (!isInteger(rows))
@@ -365,7 +365,7 @@ SEXP assign(SEXP dt, SEXP rows, SEXP cols, SEXP newcolnames, SEXP values)
     }
     if (k>0) {
       if (!isDataTable) error("set() on a data.frame is for changing existing columns, not adding new ones. Please use a data.table for that. data.table's are over-allocated and don't shallow copy.");
-      PROTECT(newcolnames = allocVector(STRSXP, k)); protecti++;
+      newcolnames = PROTECT(allocVector(STRSXP, k)); protecti++;
       for (i=0; i<k; i++) {
         SET_STRING_ELT(newcolnames, i, STRING_ELT(cols, buf[i]));
         INTEGER(tmp)[buf[i]] = oldncol+i+1;
@@ -467,6 +467,7 @@ SEXP assign(SEXP dt, SEXP rows, SEXP cols, SEXP newcolnames, SEXP values)
     // truelengths of both already set by alloccol
   }
   for (i=0; i<length(cols); i++) {
+    int thisprotecti = 0;  // UNPROTECT(thisprotecti) at the end of this loop to save protection stack
     coln = INTEGER(cols)[i]-1;
     SEXP thisvalue = RHS_list_of_columns ? VECTOR_ELT(values, i) : values;
     if (TYPEOF(thisvalue)==NILSXP) {
@@ -488,18 +489,19 @@ SEXP assign(SEXP dt, SEXP rows, SEXP cols, SEXP newcolnames, SEXP values)
       } else {
         if (verbose) Rprintf("Direct plonk of unnamed RHS, no copy.\n");  // e.g. DT[,a:=as.character(a)] as tested by 754.3
       }
-      SET_VECTOR_ELT(dt,coln,thisvalue);                   // plonk new column in as it's already the correct length
+      SET_VECTOR_ELT(dt, coln, thisvalue);                 // plonk new column in as it's already the correct length
       setAttrib(thisvalue, R_NamesSymbol, R_NilValue);     // clear names such as  DT[,a:=mapvector[a]]
       setAttrib(thisvalue, R_DimSymbol, R_NilValue);       // so that matrix is treated as vector
       setAttrib(thisvalue, R_DimNamesSymbol, R_NilValue);  // the 3rd of the 3 attribs not copied by copyMostAttrib, for consistency.
       continue;
     }
+    SEXP RHS;
+
     if (coln+1 > oldncol) {  // new column
-      newcol = allocNAVector(TYPEOF(thisvalue),nrow);
+      SET_VECTOR_ELT(dt, coln, newcol=allocNAVector(TYPEOF(thisvalue), nrow));
       // initialize with NAs for when 'rows' is a subset and it doesn't touch
       // do not try to save the time to NA fill (contiguous branch free assign anyway) since being
       // sure all items will be written to (isNull(rows), length(rows), vlen<1, targetlen) is not worth the risk.
-      SET_VECTOR_ELT(dt,coln,newcol);
       if (isVectorAtomic(thisvalue)) copyMostAttrib(thisvalue,newcol);  // class etc but not names
       // else for lists (such as data.frame and data.table) treat them as raw lists and drop attribs
       if (vlen<1) continue;   // e.g. DT[,newcol:=integer()] (adding new empty column)
@@ -512,7 +514,7 @@ SEXP assign(SEXP dt, SEXP rows, SEXP cols, SEXP newcolnames, SEXP values)
         // If it's the same RHS being assigned to several columns, we have to recoerce for each
         // one because the levels of each target are likely different
         if (isFactor(thisvalue)) {
-          PROTECT(thisvalue = asCharacterFactor(thisvalue)); protecti++;
+          thisvalue = PROTECT(asCharacterFactor(thisvalue)); thisprotecti++;
         }
         targetlevels = getAttrib(targetcol, R_LevelsSymbol);
         if (isNull(targetlevels)) error("somehow this factor column has no levels");
@@ -535,15 +537,15 @@ SEXP assign(SEXP dt, SEXP rows, SEXP cols, SEXP newcolnames, SEXP values)
           }
           R_len_t addi = 0;
           SEXP addlevels=NULL;
-          PROTECT(RHS = allocVector(INTSXP, length(thisvalue))); protecti++;
+          RHS = PROTECT(allocVector(INTSXP, length(thisvalue))); thisprotecti++;
           int *iRHS = INTEGER(RHS);
           for (j=0; j<length(thisvalue); j++) {
             thisv = STRING_ELT(thisvalue,j);
             if (TRUELENGTH(thisv)==0) {
               if (addi==0) {
-                PROTECT(addlevels = allocVector(STRSXP, 100)); protecti++;
+                addlevels = PROTECT(allocVector(STRSXP, 100)); thisprotecti++;
               } else if (addi >= length(addlevels)) {
-                PROTECT(addlevels = growVector(addlevels, length(addlevels)+1000)); protecti++;
+                addlevels = PROTECT(growVector(addlevels, length(addlevels)+1000)); thisprotecti++;
               }
               SET_STRING_ELT(addlevels,addi++,thisv);
               // if-else for #1718 fix
@@ -553,7 +555,7 @@ SEXP assign(SEXP dt, SEXP rows, SEXP cols, SEXP newcolnames, SEXP values)
           }
           if (addi > 0) {
             R_len_t oldlen = length(targetlevels);
-            PROTECT(targetlevels = growVector(targetlevels, oldlen+addi)); protecti++;
+            targetlevels = PROTECT(growVector(targetlevels, oldlen+addi)); thisprotecti++;
             for (j=0; j<addi; j++)
               SET_STRING_ELT(targetlevels, oldlen+j, STRING_ELT(addlevels, j));
             setAttrib(targetcol, R_LevelsSymbol, targetlevels);
@@ -566,13 +568,13 @@ SEXP assign(SEXP dt, SEXP rows, SEXP cols, SEXP newcolnames, SEXP values)
             error("Internal error: up front checks (before starting to modify DT) didn't catch type of RHS ('%s') assigning to factor column '%s'. please report to data.table issue tracker.", type2char(TYPEOF(thisvalue)), CHAR(STRING_ELT(names,coln))); // # nocov
           int *iRHS;
           if (isReal(thisvalue) || TYPEOF(thisvalue)==LGLSXP) {
-            PROTECT(RHS = coerceVector(thisvalue,INTSXP)); protecti++;
+            RHS = PROTECT(coerceVector(thisvalue,INTSXP)); thisprotecti++;
             iRHS = INTEGER(RHS);
             // silence warning on singleton NAs
             if (iRHS[0] != NA_INTEGER) warning("Coerced '%s' RHS to 'integer' to match the factor column's underlying type. Character columns are now recommended (can be in keys), or coerce RHS to integer or character first.", type2char(TYPEOF(thisvalue)));
           } else { // thisvalue is integer
             // make sure to copy thisvalue. May be modified below. See #2984
-            RHS = PROTECT(duplicate(thisvalue)); protecti++;
+            RHS = PROTECT(duplicate(thisvalue)); thisprotecti++;
             iRHS = INTEGER(RHS);
           }
           for (int j=0; j<length(RHS); j++) {
@@ -589,11 +591,11 @@ SEXP assign(SEXP dt, SEXP rows, SEXP cols, SEXP newcolnames, SEXP values)
         else {
           // coerce the RHS to match the type of the column, unlike [<-.data.frame, for efficiency.
           if (isString(targetcol) && isFactor(thisvalue)) {
-            PROTECT(RHS = asCharacterFactor(thisvalue)); protecti++;
+            RHS = PROTECT(asCharacterFactor(thisvalue)); thisprotecti++;
             if (verbose) Rprintf("Coerced factor RHS to character to match the column's type. Avoid this coercion if possible, for efficiency, by creating RHS as type character.\n");
             // TO DO: datatable.pedantic could turn this into warning
           } else {
-            PROTECT(RHS = coerceVector(thisvalue,TYPEOF(targetcol))); protecti++;
+            RHS = PROTECT(coerceVector(thisvalue,TYPEOF(targetcol))); thisprotecti++;
             char *s1 = (char *)type2char(TYPEOF(targetcol));
             char *s2 = (char *)type2char(TYPEOF(thisvalue));
             // FR #2551, added test for equality between RHS and thisvalue to not provide the warning when length(thisvalue) == 1
@@ -623,9 +625,10 @@ SEXP assign(SEXP dt, SEXP rows, SEXP cols, SEXP newcolnames, SEXP values)
       }
     }
     memrecycle(targetcol, rows, 0, targetlen, RHS);  // also called from dogroups where these arguments are used more
+    UNPROTECT(thisprotecti); // unprotect inside loop through columns to save protection stack
   }
   *_Last_updated = numToDo;  // the updates have taken place with no error, so update .Last.updated now
-  PROTECT(assignedNames = allocVector(STRSXP, LENGTH(cols))); protecti++;
+  assignedNames = PROTECT(allocVector(STRSXP, LENGTH(cols))); protecti++;
   for (i=0;i<LENGTH(cols);i++) SET_STRING_ELT(assignedNames,i,STRING_ELT(names,INTEGER(cols)[i]-1));
   key = getAttrib(dt, sym_sorted);
   if (length(key)) {
@@ -1124,7 +1127,8 @@ SEXP setcolorder(SEXP x, SEXP o)
   }
   Free(seen);
 
-  SEXP *tmp = Calloc(ncol, SEXP), *xd = VECTOR_PTR(x), *namesd = STRING_PTR(names);
+  SEXP *tmp = Calloc(ncol, SEXP);
+  SEXP *xd = VECTOR_PTR(x), *namesd = STRING_PTR(names);
   for (int i=0; i<ncol; ++i) tmp[i] = xd[od[i]-1];
   memcpy(xd, tmp, ncol*sizeof(SEXP)); // sizeof is type size_t so no overflow here
   for (int i=0; i<ncol; ++i) tmp[i] = namesd[od[i]-1];

--- a/src/assign.c
+++ b/src/assign.c
@@ -279,37 +279,27 @@ SEXP assign(SEXP dt, SEXP rows, SEXP cols, SEXP newcolnames, SEXP values)
   // cols : column names or numbers corresponding to the values to set
   // rows : row numbers to assign
   R_len_t i, j, numToDo, targetlen, vlen, r, oldncol, oldtncol, coln, protecti=0, newcolnum, indexLength;
-  SEXP targetcol, names, nullint, thisv, targetlevels, newcol, s, colnam, klass, tmp, colorder, key, index, a, assignedNames, indexNames;
+  SEXP targetcol, names, nullint, thisv, targetlevels, newcol, s, colnam, tmp, colorder, key, index, a, assignedNames, indexNames;
   SEXP bindingIsLocked = getAttrib(dt, install(".data.table.locked"));
-  bool verbose=GetVerbose(), anytodelete=false, isDataTable=false;
+  bool verbose=GetVerbose(), anytodelete=false;
   const char *c1, *tc1, *tc2;
-  int *buf, k=0, newKeyLength, indexNo;
+  int *buf, newKeyLength, indexNo;
   size_t size; // must be size_t otherwise overflow later in memcpy
   if (isNull(dt)) error("assign has been passed a NULL dt");
   if (TYPEOF(dt) != VECSXP) error("dt passed to assign isn't type VECSXP");
   if (length(bindingIsLocked) && LOGICAL(bindingIsLocked)[0])
     error(".SD is locked. Updating .SD by reference using := or set are reserved for future use. Use := in j directly. Or use copy(.SD) as a (slow) last resort, until shallow() is exported.");
 
-  klass = getAttrib(dt, R_ClassSymbol);
-  if (isNull(klass)) error("Input passed to assign has no class attribute. Must be a data.table or data.frame.");
-  // Check if there is a class "data.table" somewhere (#5115).
   // We allow set() on data.frame too; e.g. package Causata uses set() on a data.frame in tests/testTransformationReplay.R
   // := is only allowed on a data.table. However, the ":=" = stop(...) message in data.table.R will have already
   // detected use on a data.frame before getting to this point.
-  for (i=0; i<length(klass); i++) {   // There doesn't seem to be an R API interface to inherits(), but manually here isn't too bad.
-    if (strcmp(CHAR(STRING_ELT(klass, i)), "data.table") == 0) break;
-  }
-  if (i<length(klass))
-    isDataTable = true;
-  else {
-    for (i=0; i<length(klass); i++) {
-      if (strcmp(CHAR(STRING_ELT(klass, i)), "data.frame") == 0) break;
-    }
-    if (i == length(klass)) error("Input is not a data.table, data.frame or an object that inherits from either.");
-    isDataTable = false;   // meaning data.frame from now on. Can use set() on existing columns but not add new ones because DF aren't over-allocated.
-  }
+  // For data.frame, can use set() on existing columns but not add new ones because DF are not over-allocated.
+  bool isDataTable = INHERITS(dt, char_datatable);
+  if (!isDataTable && !INHERITS(dt, char_dataframe))
+    error("Input is not a data.table, data.frame or an object that inherits from either.");
+
   oldncol = LENGTH(dt);
-  names = getAttrib(dt,R_NamesSymbol);
+  names = getAttrib(dt, R_NamesSymbol);
   if (isNull(names)) error("dt passed to assign has no names");
   if (length(names)!=oldncol)
     error("Internal error in assign: length of names (%d) is not length of dt (%d)",length(names),oldncol); // # nocov
@@ -360,6 +350,7 @@ SEXP assign(SEXP dt, SEXP rows, SEXP cols, SEXP newcolnames, SEXP values)
   if (isString(cols)) {
     PROTECT(tmp = chmatch(cols, names, 0)); protecti++;
     buf = (int *) R_alloc(length(cols), sizeof(int));
+    int k=0;
     for (i=0; i<length(cols); i++) {
       if (INTEGER(tmp)[i] == 0) buf[k++] = i;
     }

--- a/src/assign.c
+++ b/src/assign.c
@@ -296,7 +296,7 @@ SEXP assign(SEXP dt, SEXP rows, SEXP cols, SEXP newcolnames, SEXP values)
   // For data.frame, can use set() on existing columns but not add new ones because DF are not over-allocated.
   bool isDataTable = INHERITS(dt, char_datatable);
   if (!isDataTable && !INHERITS(dt, char_dataframe))
-    error("Input is not a data.table, data.frame or an object that inherits from either.");
+    error("Internal error: dt passed to Cassign is not a data.table or data.frame");  // # nocov
 
   oldncol = LENGTH(dt);
   names = getAttrib(dt, R_NamesSymbol);

--- a/src/bmerge.c
+++ b/src/bmerge.c
@@ -206,9 +206,6 @@ static union {
   SEXP s;
 } ival, xval;
 
-static int mid, tmplow, tmpupp;  // global to save them being added to recursive stack. Maybe optimizer would do this anyway.
-static SEXP ic, xc;
-
 static uint64_t i64twiddle(void *p, int i)
 {
   return ((uint64_t *)p)[i] ^ 0x8000000000000000;
@@ -221,22 +218,26 @@ void bmerge_r(int xlowIn, int xuppIn, int ilowIn, int iuppIn, int col, int thisg
 // uppmax=1 if xuppIn is the upper bound of this group (needed for roll)
 // new: col starts with -1 for non-equi joins, which gathers rows from nested id group counter 'thisgrp'
 {
-  int xlow=xlowIn, xupp=xuppIn, ilow=ilowIn, iupp=iuppIn, j, k, ir, lir, tmp;
-  bool isInt64=false;
+  int xlow=xlowIn, xupp=xuppIn, ilow=ilowIn, iupp=iuppIn, j, k, ir, lir, tmp, tmplow, tmpupp;
   ir = lir = ilow + (iupp-ilow)/2;           // lir = logical i row.
   if (o) ir = o[lir]-1;                      // ir = the actual i row if i were ordered
+  SEXP ic, xc;
   if (col>-1) {
     ic = VECTOR_ELT(i,icols[col]-1);  // ic = i column
     xc = VECTOR_ELT(x,xcols[col]-1);  // xc = x column
     // it was checked in bmerge() that the types are equal
-  } else xc = nqgrp;
+  } else {
+    ic = R_NilValue;
+    xc = nqgrp;
+  }
+  bool isInt64=false;
   switch (TYPEOF(xc)) {
   case LGLSXP : case INTSXP : {  // including factors
-    const int *iic = INTEGER(ic);
+    const int *iic = (col>-1) ? INTEGER(ic) : NULL;
     const int *ixc = INTEGER(xc);
     ival.i = (col>-1) ? iic[ir] : thisgrp;
     while(xlow < xupp-1) {
-      mid = xlow + (xupp-xlow)/2;   // Same as (xlow+xupp)/2 but without risk of overflow
+      int mid = xlow + (xupp-xlow)/2;   // Same as (xlow+xupp)/2 but without risk of overflow
       xval.i = ixc[XIND(mid)];
       if (xval.i<ival.i) {          // relies on NA_INTEGER == INT_MIN, tested in init.c
         xlow=mid;
@@ -249,12 +250,12 @@ void bmerge_r(int xlowIn, int xuppIn, int ilowIn, int iuppIn, int col, int thisg
         tmplow = mid;
         tmpupp = mid;
         while(tmplow<xupp-1) {
-          mid = tmplow + (xupp-tmplow)/2;
+          int mid = tmplow + (xupp-tmplow)/2;
           xval.i = ixc[XIND(mid)];
           if (xval.i == ival.i) tmplow=mid; else xupp=mid;
         }
         while(xlow<tmpupp-1) {
-          mid = xlow + (tmpupp-xlow)/2;
+          int mid = xlow + (tmpupp-xlow)/2;
           xval.i = ixc[XIND(mid)];
           if (xval.i == ival.i) tmpupp=mid; else xlow=mid;
         }
@@ -273,7 +274,7 @@ void bmerge_r(int xlowIn, int xuppIn, int ilowIn, int iuppIn, int col, int thisg
       if (op[col] <= 3 && xlow<xupp-1 && ival.i != NA_INTEGER && INTEGER(xc)[XIND(xlow+1)] == NA_INTEGER) {
         tmplow = xlow; tmpupp = xupp;
         while (tmplow < tmpupp-1) {
-          mid = tmplow + (tmpupp-tmplow)/2;
+          int mid = tmplow + (tmpupp-tmplow)/2;
           xval.i = ixc[XIND(mid)];
           if (xval.i == NA_INTEGER) tmplow = mid; else tmpupp = mid;
         }
@@ -284,13 +285,13 @@ void bmerge_r(int xlowIn, int xuppIn, int ilowIn, int iuppIn, int col, int thisg
     tmpupp = lir;
     if (col>-1) {
       while(tmplow<iupp-1) {   // TO DO: could double up from lir rather than halving from iupp
-        mid = tmplow + (iupp-tmplow)/2;
+        int mid = tmplow + (iupp-tmplow)/2;
         xval.i = iic[ o ? o[mid]-1 : mid ];   // reuse xval to search in i
         if (xval.i == ival.i) tmplow=mid; else iupp=mid;
         // if we could guarantee ivals to be *always* sorted for all columns independently (= max(nestedid) = 1), then we can speed this up by 2x by adding checks for GE,GT,LE,LT separately.
       }
       while(ilow<tmpupp-1) {
-        mid = ilow + (tmpupp-ilow)/2;
+        int mid = ilow + (tmpupp-ilow)/2;
         xval.i = iic[ o ? o[mid]-1 : mid ];
         if (xval.i == ival.i) tmpupp=mid; else ilow=mid;
       }
@@ -302,19 +303,19 @@ void bmerge_r(int xlowIn, int xuppIn, int ilowIn, int iuppIn, int col, int thisg
     if (op[col] != EQ) error("Only '==' operator is supported for columns of type %s.", type2char(TYPEOF(xc)));
     ival.s = ENC2UTF8(STRING_ELT(ic,ir));
     while(xlow < xupp-1) {
-      mid = xlow + (xupp-xlow)/2;
+      int mid = xlow + (xupp-xlow)/2;
       xval.s = ENC2UTF8(STRING_ELT(xc, XIND(mid)));
       tmp = StrCmp(xval.s, ival.s);  // uses pointer equality first, NA_STRING are allowed and joined to, then uses strcmp on CHAR().
       if (tmp == 0) {                // TO DO: deal with mixed encodings and locale optionally
         tmplow = mid;
         tmpupp = mid;
         while(tmplow<xupp-1) {
-          mid = tmplow + (xupp-tmplow)/2;
+          int mid = tmplow + (xupp-tmplow)/2;
           xval.s = ENC2UTF8(STRING_ELT(xc, XIND(mid)));
           if (ival.s == xval.s) tmplow=mid; else xupp=mid;  // the == here handles encodings as well. Marked non-utf8 encodings are converted to utf-8 using ENC2UTF8.
         }
         while(xlow<tmpupp-1) {
-          mid = xlow + (tmpupp-xlow)/2;
+          int mid = xlow + (tmpupp-xlow)/2;
           xval.s = ENC2UTF8(STRING_ELT(xc, XIND(mid)));
           if (ival.s == xval.s) tmpupp=mid; else xlow=mid;  // see above re ==
         }
@@ -328,12 +329,12 @@ void bmerge_r(int xlowIn, int xuppIn, int ilowIn, int iuppIn, int col, int thisg
     tmplow = lir;
     tmpupp = lir;
     while(tmplow<iupp-1) {
-      mid = tmplow + (iupp-tmplow)/2;
+      int mid = tmplow + (iupp-tmplow)/2;
       xval.s = ENC2UTF8(STRING_ELT(ic, o ? o[mid]-1 : mid));
       if (xval.s == ival.s) tmplow=mid; else iupp=mid;   // see above re ==
     }
     while(ilow<tmpupp-1) {
-      mid = ilow + (tmpupp-ilow)/2;
+      int mid = ilow + (tmpupp-ilow)/2;
       xval.s = ENC2UTF8(STRING_ELT(ic, o ? o[mid]-1 : mid));
       if (xval.s == ival.s) tmpupp=mid; else ilow=mid;   // see above re ==
     }
@@ -346,7 +347,7 @@ void bmerge_r(int xlowIn, int xuppIn, int ilowIn, int iuppIn, int col, int thisg
     // TODO: remove this last remaining use of i64twiddle.
     ival.ull = twiddle(dic, ir);
     while(xlow < xupp-1) {
-      mid = xlow + (xupp-xlow)/2;
+      int mid = xlow + (xupp-xlow)/2;
       xval.ull = twiddle(dxc, XIND(mid));
       if (xval.ull<ival.ull) {
         xlow=mid;
@@ -356,12 +357,12 @@ void bmerge_r(int xlowIn, int xuppIn, int ilowIn, int iuppIn, int col, int thisg
         tmplow = mid;
         tmpupp = mid;
         while(tmplow<xupp-1) {
-          mid = tmplow + (xupp-tmplow)/2;
+          int mid = tmplow + (xupp-tmplow)/2;
           xval.ull = twiddle(dxc, XIND(mid));
           if (xval.ull == ival.ull) tmplow=mid; else xupp=mid;
         }
         while(xlow<tmpupp-1) {
-          mid = xlow + (tmpupp-xlow)/2;
+          int mid = xlow + (tmpupp-xlow)/2;
           xval.ull = twiddle(dxc, XIND(mid));
           if (xval.ull == ival.ull) tmpupp=mid; else xlow=mid;
         }
@@ -380,7 +381,7 @@ void bmerge_r(int xlowIn, int xuppIn, int ilowIn, int iuppIn, int col, int thisg
       if (op[col] <= 3 && xlow<xupp-1 && !isivalNA && (!isInt64 ? ISNAN(dxc[XIND(xlow+1)]) : (DtoLL(dxc[XIND(xlow+1)]) == NA_INT64_LL))) {
         tmplow = xlow; tmpupp = xupp;
         while (tmplow < tmpupp-1) {
-          mid = tmplow + (tmpupp-tmplow)/2;
+          int mid = tmplow + (tmpupp-tmplow)/2;
           xval.d = REAL(xc)[XIND(mid)];
           if (!isInt64 ? ISNAN(xval.d) : xval.ll == NA_INT64_LL) tmplow = mid; else tmpupp = mid;
         }
@@ -391,12 +392,12 @@ void bmerge_r(int xlowIn, int xuppIn, int ilowIn, int iuppIn, int col, int thisg
     tmpupp = lir;
     if (col>-1) {
       while(tmplow<iupp-1) {
-        mid = tmplow + (iupp-tmplow)/2;
+        int mid = tmplow + (iupp-tmplow)/2;
         xval.ull = twiddle(dic, o ? o[mid]-1 : mid);
         if (xval.ull == ival.ull) tmplow=mid; else iupp=mid;
       }
       while(ilow<tmpupp-1) {
-        mid = ilow + (tmpupp-ilow)/2;
+        int mid = ilow + (tmpupp-ilow)/2;
         xval.ull = twiddle(dic, o ? o[mid]-1 : mid);
         if (xval.ull == ival.ull) tmpupp=mid; else ilow=mid;
       }

--- a/src/data.table.h
+++ b/src/data.table.h
@@ -75,6 +75,7 @@ SEXP char_allLen1;
 SEXP char_allGrp1;
 SEXP char_factor;
 SEXP char_ordered;
+SEXP char_datatable;
 SEXP char_dataframe;
 SEXP char_NULL;
 SEXP sym_sorted;

--- a/src/fmelt.c
+++ b/src/fmelt.c
@@ -380,12 +380,8 @@ static SEXP combineFactorLevels(SEXP factorLevels, SEXP target, int * factorType
 }
 
 SEXP getvaluecols(SEXP DT, SEXP dtnames, Rboolean valfactor, Rboolean verbose, struct processData *data) {
-  int i, j, k, counter=0, thislen=0;
-  SEXP thisvaluecols, ansvals, thisidx=R_NilValue, flevels;
-  Rboolean copyattr = FALSE, thisvalfactor;
-  size_t size;
-  for (i=0; i<data->lvalues; i++) {
-    thisvaluecols = VECTOR_ELT(data->valuecols, i);
+  for (int i=0; i<data->lvalues; ++i) {
+    SEXP thisvaluecols = VECTOR_ELT(data->valuecols, i);
     if (!data->isidentical[i])
       warning("'measure.vars' [%s] are not all of the same type. By order of hierarchy, the molten data value column will be of type '%s'. All measure variables not of type '%s' will be coerced too. Check DETAILS in ?melt.data.table for more on coercion.\n", CHAR(STRING_ELT(concat(dtnames, thisvaluecols), 0)), type2char(data->maxtype[i]), type2char(data->maxtype[i]));
     if (data->maxtype[i] == VECSXP && data->narm) {
@@ -395,11 +391,11 @@ SEXP getvaluecols(SEXP DT, SEXP dtnames, Rboolean valfactor, Rboolean verbose, s
   }
   if (data->narm) {
     SEXP seqcols = PROTECT(seq_int(data->lvalues, 1));
-    for (i=0; i<data->lmax; i++) {
+    for (int i=0; i<data->lmax; ++i) {
       SEXP tmp = PROTECT(allocVector(VECSXP, data->lvalues));
-      for (j=0; j<data->lvalues; j++) {
+      for (int j=0; j<data->lvalues; ++j) {
         if (i < data->leach[j]) {
-          thisvaluecols = VECTOR_ELT(data->valuecols, j);
+          SEXP thisvaluecols = VECTOR_ELT(data->valuecols, j);
           SET_VECTOR_ELT(tmp, j, VECTOR_ELT(DT, INTEGER(thisvaluecols)[i]-1));
         } else {
           SET_VECTOR_ELT(tmp, j, allocNAVector(data->maxtype[j], data->nrow));
@@ -415,34 +411,36 @@ SEXP getvaluecols(SEXP DT, SEXP dtnames, Rboolean valfactor, Rboolean verbose, s
   } else {
     data->totlen = data->nrow * data->lmax;
   }
-  flevels = PROTECT(allocVector(VECSXP, data->lmax));
+  SEXP flevels = PROTECT(allocVector(VECSXP, data->lmax));
   Rboolean *isordered = (Rboolean *)R_alloc(data->lmax, sizeof(Rboolean));
-  ansvals = PROTECT(allocVector(VECSXP, data->lvalues));
-  for (i=0; i<data->lvalues; i++) {
-    thisvalfactor = (data->maxtype[i] == VECSXP) ? FALSE : valfactor;
+  SEXP ansvals = PROTECT(allocVector(VECSXP, data->lvalues));
+  for (int i=0; i<data->lvalues; ++i) {
+    bool thisvalfactor = (data->maxtype[i] == VECSXP) ? false : valfactor;
     SEXP target = PROTECT(allocVector(data->maxtype[i], data->totlen)); // to keep rchk happy
     SET_VECTOR_ELT(ansvals, i, target);
     UNPROTECT(1);  // still protected by virtue of being member of protected ansval.
-    thisvaluecols = VECTOR_ELT(data->valuecols, i);
-    counter = 0; copyattr = FALSE;
-    for (j=0; j<data->lmax; j++) {
+    SEXP thisvaluecols = VECTOR_ELT(data->valuecols, i);
+    int counter = 0;
+    bool copyattr = false;
+    for (int j=0; j<data->lmax; ++j) {
       int thisprotecti = 0;
       SEXP thiscol = (j < data->leach[i]) ? VECTOR_ELT(DT, INTEGER(thisvaluecols)[j]-1)
                        : allocNAVector(data->maxtype[i], data->nrow);
       if (!copyattr && data->isidentical[i] && !data->isfactor[i]) {
         copyMostAttrib(thiscol, target);
-        copyattr = TRUE;
+        copyattr = true;
       }
       if (TYPEOF(thiscol) != TYPEOF(target) && (data->maxtype[i] == VECSXP || !isFactor(thiscol))) {
         thiscol = PROTECT(coerceVector(thiscol, TYPEOF(target)));  thisprotecti++;
       }
-      const int *ithisidx;
+      const int *ithisidx = NULL;
+      int thislen = 0;
       if (data->narm) {
-        thisidx = VECTOR_ELT(data->naidx, j);
+        SEXP thisidx = VECTOR_ELT(data->naidx, j);
         ithisidx = INTEGER(thisidx);
         thislen = length(thisidx);
       }
-      size = SIZEOF(thiscol);
+      size_t size = SIZEOF(thiscol);
       switch (TYPEOF(target)) {
       case VECSXP :
         if (data->narm) {
@@ -461,17 +459,17 @@ SEXP getvaluecols(SEXP DT, SEXP dtnames, Rboolean valfactor, Rboolean verbose, s
           } else SET_VECTOR_ELT(flevels, j, thiscol);
         }
         if (data->narm) {
-          for (k=0; k<thislen; k++)
+          for (int k=0; k<thislen; ++k)
             SET_STRING_ELT(target, counter + k, STRING_ELT(thiscol, ithisidx[k]-1));
         } else {
-          for (k=0; k<data->nrow; k++) SET_STRING_ELT(target, j*data->nrow + k, STRING_ELT(thiscol, k));
+          for (int k=0; k<data->nrow; ++k) SET_STRING_ELT(target, j*data->nrow + k, STRING_ELT(thiscol, k));
         }
         break;
       case REALSXP : {
         double *dtarget = REAL(target);
         const double *dthiscol = REAL(thiscol);
         if (data->narm) {
-          for (k=0; k<thislen; k++)
+          for (int k=0; k<thislen; ++k)
             dtarget[counter + k] = dthiscol[ithisidx[k]-1];
         } else {
           memcpy(dtarget + j*data->nrow, dthiscol, data->nrow*size);

--- a/src/freadR.c
+++ b/src/freadR.c
@@ -216,8 +216,7 @@ SEXP freadR(
 
 static void applyDrop(SEXP items, int8_t *type, int ncol, int dropSource) {
   if (!length(items)) return;
-  SEXP itemsInt = PROTECT(isString(items) ? chmatch(items, colNamesSxp, NA_INTEGER) :
-                                            coerceVector(items, INTSXP));
+  SEXP itemsInt = PROTECT( isString(items) ? chmatch(items, colNamesSxp, NA_INTEGER) : coerceVector(items, INTSXP) );
   const int *itemsD = INTEGER(itemsInt), n=LENGTH(itemsInt);
   for (int j=0; j<n; ++j) {
     int k = itemsD[j];

--- a/src/init.c
+++ b/src/init.c
@@ -271,6 +271,7 @@ void attribute_visible R_init_datatable(DllInfo *info)
   char_allGrp1 =   PRINTNAME(install("allGrp1"));
   char_factor =    PRINTNAME(install("factor"));
   char_ordered =   PRINTNAME(install("ordered"));
+  char_datatable = PRINTNAME(install("data.table"));
   char_dataframe = PRINTNAME(install("data.frame"));
   char_NULL =      PRINTNAME(install("NULL"));
 

--- a/src/rbindlist.c
+++ b/src/rbindlist.c
@@ -305,12 +305,13 @@ SEXP rbindlist(SEXP l, SEXP usenamesArg, SEXP fillArg, SEXP idcolArg)
       }
       if (firsti==-1) { firsti=i; firstw=w; firstCol=thisCol; }
       else {
-        if (!factor && !int64 && !R_compute_identical(PROTECT(getAttrib(thisCol, R_ClassSymbol)),
-                                                      PROTECT(getAttrib(firstCol, R_ClassSymbol)),
-                                                      0)) {
+        bool prot = false;
+        if (!factor && !int64 && (prot=true) && !R_compute_identical(PROTECT(getAttrib(thisCol, R_ClassSymbol)),
+                                                                     PROTECT(getAttrib(firstCol, R_ClassSymbol)),
+                                                                     0)) {
           error("Class attribute on column %d of item %d does not match with column %d of item %d.", w+1, i+1, firstw+1, firsti+1);
         }
-        UNPROTECT(2);
+        if (prot) UNPROTECT(2);
       }
     }
 

--- a/src/rbindlist.c
+++ b/src/rbindlist.c
@@ -304,8 +304,13 @@ SEXP rbindlist(SEXP l, SEXP usenamesArg, SEXP fillArg, SEXP idcolArg)
         int64=true;
       }
       if (firsti==-1) { firsti=i; firstw=w; firstCol=thisCol; }
-      else if (!factor && !int64 && !R_compute_identical(getAttrib(thisCol, R_ClassSymbol), getAttrib(firstCol,R_ClassSymbol), 0)) {
-        error("Class attribute on column %d of item %d does not match with column %d of item %d.", w+1, i+1, firstw+1, firsti+1);
+      else {
+        if (!factor && !int64 && !R_compute_identical(PROTECT(getAttrib(thisCol, R_ClassSymbol)),
+                                                      PROTECT(getAttrib(firstCol, R_ClassSymbol)),
+                                                      0)) {
+          error("Class attribute on column %d of item %d does not match with column %d of item %d.", w+1, i+1, firstw+1, firsti+1);
+        }
+        UNPROTECT(2);
       }
     }
 

--- a/src/transpose.c
+++ b/src/transpose.c
@@ -4,11 +4,6 @@
 
 SEXP transpose(SEXP l, SEXP fill, SEXP ignoreArg) {
 
-  R_len_t k=0, maxlen=0, zerolen=0, anslen;
-  SEXP li, ans;
-  SEXPTYPE type, maxtype=0;
-  Rboolean coerce = FALSE;
-
   if (!isNewList(l))
     error("l must be a list.");
   if (!length(l))
@@ -22,8 +17,10 @@ SEXP transpose(SEXP l, SEXP fill, SEXP ignoreArg) {
 
   // preprocessing
   R_len_t *len  = (R_len_t *)R_alloc(ln, sizeof(R_len_t));
+  int maxlen=0, zerolen=0;
+  SEXPTYPE maxtype=0;
   for (int i=0; i<ln; ++i) {
-    li = VECTOR_ELT(l, i);
+    SEXP li = VECTOR_ELT(l, i);
     if (!isVectorAtomic(li) && !isNull(li))
       error("Item %d of list input is not an atomic vector", i+1);
     len[i] = length(li);
@@ -33,7 +30,7 @@ SEXP transpose(SEXP l, SEXP fill, SEXP ignoreArg) {
     if (isFactor(li)) {
       maxtype = STRSXP;
     } else {
-      type = TYPEOF(li);
+      SEXPTYPE type = TYPEOF(li);
       if (type > maxtype)
         maxtype = type;
     }
@@ -42,20 +39,21 @@ SEXP transpose(SEXP l, SEXP fill, SEXP ignoreArg) {
   fill = PROTECT(coerceVector(fill, maxtype));
 
   // allocate 'ans'
-  ans = PROTECT(allocVector(VECSXP, maxlen));
-  anslen = (!ignore) ? ln : (ln - zerolen);
+  SEXP ans = PROTECT(allocVector(VECSXP, maxlen));
+  int anslen = (!ignore) ? ln : (ln - zerolen);
   for (int i=0; i<maxlen; ++i) {
     SET_VECTOR_ELT(ans, i, allocVector(maxtype, anslen));
   }
 
   // transpose
+  int k=0;
   for (int i=0; i<ln; ++i) {
     if (ignore && !len[i]) continue;
-    li = VECTOR_ELT(l, i);
+    SEXP li = VECTOR_ELT(l, i);
+    bool coerce = false;
     if (TYPEOF(li) != maxtype) {
-      coerce = TRUE;
-      if (!isFactor(li)) li = PROTECT(coerceVector(li, maxtype));
-      else li = PROTECT(asCharacterFactor(li));
+      coerce = true;
+      li = PROTECT(isFactor(li) ? asCharacterFactor(li) : coerceVector(li, maxtype));
     }
     switch (maxtype) { // TODO remove more macros
     case INTSXP : {
@@ -94,10 +92,7 @@ SEXP transpose(SEXP l, SEXP fill, SEXP ignoreArg) {
     default :
         error("Unsupported column type '%s'", type2char(maxtype));
     }
-    if (coerce) {
-      coerce = FALSE;
-      UNPROTECT(1);
-    }
+    if (coerce) UNPROTECT(1);
     k++;
   }
   UNPROTECT(2);

